R/main.R

```diff
@@ -114,68 +134,96 @@ Options:
 
 Parameters, if given, must be passed through in key=value pairs"
 
-
-main_do_run <- function(x) {
-  ## TODO: Get some classed errors though here and then write out
-  ## information about whether or not things worked and why they
-  ## didn't.  Possible issues (in order)
-  ##
-  ## * orderly report not found
-  ## * error while preparing (e.g., package not found)
-  ## * error while running report
-  ## * error checking artefacts
-  config <- orderly_config_get(x$options$root, TRUE)
-  name <- x$options$name
-  commit <- !x$options$no_commit
-  instance <- x$options$instance
-  parameters <- x$options$parameters
-  id_file <- x$options$id_file
-  parameters <- x$options$parameters
-  print_log <- x$options$print_log
-  ref <- x$options$ref
-  fetch <- x$options$fetch
-  pull <- x$options$pull
-  message <- x$options$message
-
-  main_run <- function() {
-    if (pull) {
-      if (is.null(ref)) {
-        git_pull(config$root)
-      } else {
-        orderly_cli_error(
-          "Can't use --pull with --ref; perhaps you meant --fetch ?")
+do_run <- function(run_func) {
+  function(x) {
+    ## TODO: Get some classed errors though here and then write out
+    ## information about whether or not things worked and why they
+    ## didn't.  Possible issues (in order)
+    ##
+    ## * orderly report not found
+    ## * error while preparing (e.g., package not found)
+    ## * error while running report
+    ## * error checking artefacts
+    config <- orderly_config_get(x$options$root, TRUE)
+    name <- x$options$name
+    commit <- !x$options$no_commit
+    instance <- x$options$instance
+    parameters <- x$options$parameters
+    id_file <- x$options$id_file
+    parameters <- x$options$parameters
+    print_log <- x$options$print_log
+    ref <- x$options$ref
+    fetch <- x$options$fetch
+    pull <- x$options$pull
+    message <- x$options$message
+
+    main_batch <- function() {
```

Comment:
This seems more complicated than ideal - I see you're trying to avoid repetition but  I think that comes at a cost of readability.  I think it's worth considering if we should be exposing all the same arguments to batch anyway - I think that id_file, commit should definitely disappear.  Some of the run complication exists to support the runner and we might want to look at that more holistically later.  Keeping the new part clean will help unpick that.